cpp/mrc/src/public/coroutines/task_container.cpp

```diff
@@ -53,20 +54,26 @@ auto TaskContainer::start(Task<void>&& user_task, GarbageCollectPolicy cleanup)
 {
     m_size.fetch_add(1, std::memory_order::relaxed);
 
-    std::scoped_lock lk{m_mutex};
-
-    if (cleanup == GarbageCollectPolicy::yes)
     {
-        gc_internal();
-    }
+        std::scoped_lock lk{m_mutex};
 
-    // Store the task inside a cleanup task for self deletion.
-    auto pos  = m_tasks.emplace(m_tasks.end(), std::nullopt);
-    auto task = make_cleanup_task(std::move(user_task), pos);
-    *pos      = std::move(task);
+        if (cleanup == GarbageCollectPolicy::yes)
+        {
+            gc_internal();
+        }
 
-    // Start executing from the cleanup task to schedule the user's task onto the thread pool.
-    pos->value().resume();
+        // Store the task inside a cleanup task for self deletion.
+        auto pos  = m_tasks.emplace(m_tasks.end(), std::nullopt);
+        auto task = make_cleanup_task(std::move(user_task), pos);
+        *pos      = std::move(task);
+
+        m_next_tasks.push(pos);
+    }
+
+    if (m_max_simultaneous_tasks <= 0 or m_size <= m_max_simultaneous_tasks)
```

Comment:
`m_max_simultaneous_tasks` gets initialized by a `size_t` which is unsigned so it can never be negative.